File: 프로그래머스/2/87946. 피로도/피로도.js
// DFS+백트래킹
// DFS로 경로 탐색 (DFS 내에서 매번 for문으로 경로를 생성함.)
// 해당 경로 방문 가능 여부 확인 후 불가능하면 백트래킹, 가능하면 DFS 마저 진행
function solution(k, dungeons) {
    let maxCount = 0; // 탐험한 최대 던전 수
    const visited = new Array(dungeons.length).fill(false);
    
    const dfs = (currPiro, count) => {
        maxCount = Math.max(maxCount, count);
        
        for (let i = 0; i < dungeons.length; i++) {
            // 탐험 불가
            if (visited[i] || currPiro < dungeons[i][0]) {
                continue;
            }
            
            visited[i] = true;
            dfs(currPiro -  dungeons[i][1], count + 1);
            visited[i] = false;
        }
    }
    
    dfs(k, maxCount);
    
    return maxCount;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    // 클래스 내의 모든 인스턴스에서 공유하기 위해 static 키워드 사용
    static int answer = 0;
    static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length]; // 모든 요소 false로 초기화
        DFS(k, 0, dungeons);
        return answer;
    }
    
    // 반환하지 않으므로 void형 사용
    public void DFS(int piro, int cnt, int[][] dungeons) {
        answer = Math.max(cnt, answer);
        
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && piro >= dungeons[i][0]) {
                visited[i] = true;
                DFS(piro - dungeons[i][1], cnt + 1, dungeons);
                visited[i] = false;
            }
        }
    }
}